cmd/test.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+const warnMessage = `WARNING: Using the host driver runs tests directly on the machine 
+that this binary is being run on, and can potentially corrupt your system.
+Be sure you know what you're doing before continuing!
+
+Continue? (y/n)`
+
+var totalTests int
+
+var driverImpl func(drivers.DriverConfig) (drivers.Driver, error)
+var args *drivers.DriverConfig
+
+var TestCmd = &cobra.Command{
+	Use:   "test",
+	Short: "Runs the tests",
+	Long:  `Runs the tests`,
+	Args: func(cmd *cobra.Command, _ []string) error {
+		return validateArgs()
+	},
+	Run: func(cmd *cobra.Command, _ []string) {
+		testResults := Run()
+		out := &output.OutWriter{
+			Verbose: verbose,
+			Quiet:   quiet,
+		}
+		os.Exit(out.OutputResults(testResults))
+	},
+}
+
+func validateArgs() error {
+	if driver == drivers.Host {
+		if metadata == "" {
+			return fmt.Errorf("Please provide path to image metadata file")
+		}
+		if imagePath != "" {
+			return fmt.Errorf("Cannot provide both image path and metadata file")
+		}
+	} else {
+		if imagePath == "" {
+			return fmt.Errorf("Please supply path to image or tarball to test against")
+		}
+		if metadata != "" {
+			return fmt.Errorf("Cannot provide both image path and metadata file")
+		}
+	}
+	if len(configFiles) == 0 {
+		return fmt.Errorf("Please provide at least one test config file")
+	}
+	return nil
+}
+
+func RunTests() []*unversioned.FullResult {
+	fullResults := make([]*unversioned.FullResult, 0)
+	for _, file := range configFiles {
+		tests, err := Parse(file)
+		if err != nil {
+			logrus.Fatalf("Error parsing config file: %s", err)
+		}
+		fullResults = append(fullResults, &unversioned.FullResult{
+			FileName: file,
+			Results:  tests.RunAll(),
+		})
+	}
+	return fullResults
+}
+
+func Parse(fp string) (types.StructureTest, error) {
+	testContents, err := ioutil.ReadFile(fp)
+	if err != nil {
+		return nil, err
+	}
+
+	// We first have to unmarshal to determine the schema version, then we unmarshal again
+	// to do the full parse.
+	var unmarshal types.Unmarshaller
+	var strictUnmarshal types.Unmarshaller
+	var versionHolder types.SchemaVersion
+
+	switch {
+	case strings.HasSuffix(fp, ".json"):
+		unmarshal = json.Unmarshal
+		strictUnmarshal = json.Unmarshal
+	case strings.HasSuffix(fp, ".yaml"):
+		unmarshal = yaml.Unmarshal
+		strictUnmarshal = yaml.UnmarshalStrict
+	default:
+		return nil, errors.New("Please provide valid JSON or YAML config file")
+	}
+
+	if err := unmarshal(testContents, &versionHolder); err != nil {
+		return nil, err
+	}
+
+	version := versionHolder.SchemaVersion
+	if version == "" {
+		return nil, errors.New("Please provide JSON schema version")
+	}
+
+	var st types.StructureTest
+	if schemaVersion, ok := types.SchemaVersions[version]; ok {
+		st = schemaVersion()
+	} else {
+		return nil, errors.New("Unsupported schema version: " + version)
+	}
+
+	strictUnmarshal(testContents, st)
```

Comment:
if strictUnmarshal = json.Unmarshal you're ignoring an error

Reply:
checking it now

Reply:
turns out checking this error actually causes the tests to fail....I didn't write this so I'm not totally sure why, but leaving as is for now